File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

// Time: O(n*logk). n is the total number of nodes. k is the number of lists

// method1
// struct NodeMin {
//     bool operator()(ListNode* a, ListNode* b) const {
//         return a->val > b->val;
//     }
// };

class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        // method1
        // priority_queue<ListNode*, vector<ListNode*>, NodeMin> pq;

        // method2 
        auto compare = [](ListNode* a, ListNode* b) {
            return a->val > b->val;
        };
        priority_queue<ListNode*, vector<ListNode*>, decltype(compare)> pq;

        for (int i = 0; i < lists.size(); i++) {
            if (lists[i]) {
                pq.push(lists[i]);
            }
        }

        ListNode* head = new ListNode(0);
        ListNode* prev = head;

        while (!pq.empty()) {
            ListNode* curr = pq.top();
            pq.pop();
            prev->next = curr;
            
            if (curr->next) {
                pq.push(curr->next);
            }

            prev = prev->next;
        }

        return head->next;
    }
};

File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
# class Solution:
#     def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
#         # need i as tie-breaker
#         hq = [(head.val, i, head) for i, head in enumerate(lists) if head]
#         heapq.heapify(hq)

#         dummy = ListNode()
#         curr = dummy

#         while hq:
#             _, i, head = heapq.heappop(hq)
#             curr.next = head
#             curr = curr.next
            
#             if head.next:
#                 heapq.heappush(hq, (head.next.val, i, head.next))

#         return dummy.next


class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        if not lists:
            return None
        elif len(lists) == 1:
            return lists[0]

        mid = len(lists) // 2
        left = self.mergeKLists(lists[:mid])
        right = self.mergeKLists(lists[mid:])
        return self.merge(left, right)

    def merge(self, left, right):
        dummy = ListNode()
        curr = dummy
        while left and right:
            if left.val <= right.val:
                curr.next = left
                left = left.next
            else:
                curr.next = right
                right = right.next
            curr = curr.next

        curr.next = left if left else right
        return dummy.next